doc: add documentation for CPU templates (#3659)
firecracker-microvm/firecracker: docs/cpu_templates/cpu-templates.md

# CPU templates

Firecracker allows users to customise how the vCPUs are represented
to the guest software by changing the following configuration:

- CPUID (x86_64 only)
- MSRs (Model Specific Registers, x86_64 only)
- ARM registers (aarch64 only)

A combination of the changes to the entities above is called a CPU template.

The functionality can be used when a user wants to mask a feature from
the guest. One of real world use cases for this is representing a heterogeneous
fleet (a fleet consisting of multiple CPU models) as a homogeneous fleet,
so the guests will experience a consistent feature set supported by the host.

Note: representing one CPU vendor as another CPU vendor is not supported.

Firecracker supports two types of CPU templates:

- Static CPU templates - a set of preconfigured CPU templates for users

>   to choose from;

Author: it was, but I agree it is not consistent. Removed.

- Custom CPU templates - users can create their own CPU templates.

## Static CPU templates

At the moment the following set of static CPU templates is supported:

| CPU template | CPU vendor | CPU model             |
|--------------|------------|-----------------------|
| C3           | Intel      | any                   |
| T2           | Intel      | any                   |
| T2A          | AMD        | Milan                 |
| T2CL         | Intel      | Cascade Lake or newer |
| T2S          | Intel      | any                   |
| V1N1         | ARM        | Neoverse V1           |

T2 and C3 templates are mapped as close as possible to AWS T2 and C3 instances

> in terms f CPU features.

Author: thanks, fixed.

making guests vulnerable to security attacks. Custom CPU templates must be
tested thoroughly before use in production.

In order to assist with creation and usage of CPU templates, there exists
a CPU template helper tool.
TODO: add a link to the CPU template helper tool when it is available.

### Configuring custom CPU templates

Configuration of a custom CPU template is performed via `/cpu-config` API endpoint.

An example of configuring a custom CPU template on x86_64:

```bash
curl --unix-socket /tmp/firecracker.socket -i  \
  -X PUT 'http://localhost/cpu-config' \
  -H 'Accept: application/json'            \
  -H 'Content-Type: application/json'      \
  -d '{
        "cpuid_modifiers": [
         {
            "leaf": "0x1",
           "subleaf": "0x0",
           "flags": 0,
           "modifiers": [
             {
              "register": "eax",
              "bitmap": "0bxxxx000000000011xx00011011110010"
             }
           ]
         }
       ],
       "msr_modifiers": [
         {
           "addr": "0x10a",
           "bitmap": "0b0000000000000000000000000000000000000000000000000000000000000000"
         }
       ]

>   }'

Author: indeed, don't know why it skewed. Fixed, thanks.

```

An example of configuring a custom CPU template on ARM:

```bash
curl --unix-socket /tmp/firecracker.socket -i  \
  -X PUT 'http://localhost/cpu-config' \
  -H 'Accept: application/json'            \
  -H 'Content-Type: application/json'      \
  -d '{
        "reg_modifiers": [
          {
            "addr": "0x603000000013c020",
            "bitmap": "0bxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0000xxxxxxxxxxxx0000xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
          }
        ]
  }'
```

### Custom CPU templates language schema

The full description of the custom CPU templates language can be found
[here](schema.json).

#### Expansion of contracted bitmaps

If a contracted version of a bitmap is given, for example, `0x101` where

> a 32-bit bitmap is expected, missing characters are implied to be `x`.

Author: you're right!

#### Default values for CPUID leaves that are not supported by KVM

If a custom CPU template makes modifications to a CPUID leaf that is
not reported as supported by KVM via `KVM_GET_SUPPORTED_CPUID`,

> the default register values of such a leaf are assumed to be `0`.

Reviewer: Is this behavior true? It seems to return error in this case.
https://github.com/firecracker-microvm/firecracker/blob/feature/cpu-templates/src/vmm/src/guest_config/x86_64/mod.rs#L84-L88

Author: You are probably right, it was an updated comment that I addressed. Removed.

# CPU templates

Firecracker allows users to customise how the vCPUs are represented
to the guest software by changing the following configuration:

- CPUID (x86_64 only)
- MSRs (Model Specific Registers, x86_64 only)
- ARM registers (aarch64 only)

A combination of the changes to the entities above is called a CPU template.

The functionality can be used when a user wants to mask a feature from
the guest. One of real world use cases for this is representing a heterogeneous
fleet (a fleet consisting of multiple CPU models) as a homogeneous fleet,
so the guests will experience a consistent feature set supported by the host.

Note: representing one CPU vendor as another CPU vendor is not supported.

Firecracker supports two types of CPU templates:

- Static CPU templates - a set of preconfigured CPU templates for users

>   to choose from;

Reviewer 2: maybe better: "... a set of build in CPU templates ...."?

Author: indeed, thanks.

> - Custom CPU templates - users can create their own CPU templates.

Reviewer 2: And here "CPU templates defined in json format and passed into Firecracker".

to use the T2S template.

The T2CL template is mapped to be close to Intel Cascade Lake.
It is not safe to use it on Intel CPUs older than Cascade Lake (such as Skylake).

The only AMD template is T2A. It is considered safe to be used with AMD Milan.

Intel T2CL and AMD T2A templates together aim to provide instruction set feature
parity between CPUs running them, so they can form a heterogeneous fleet
exposing the same instruction sets to the application.

The V1N1 template is designed to represent ARM Neoverse V1 as ARM Neoverse N1.

### Configuring static CPU templates

Configuration of a static CPU template is performed via `/machine-config` API endpoint:

```bash
curl --unix-socket /tmp/firecracker.socket -i  \
  -X PUT 'http://localhost/machine-config' \
  -H 'Accept: application/json'            \
  -H 'Content-Type: application/json'      \
  -d '{
           "vcpu_count": 2,
           "mem_size_mib": 1024,
           "cpu_template": "T2CL"
  }'
```

## Custom CPU templates

Users can create their own CPU templates by creating a json file containing
modifiers for CPUID, MSRs or ARM registers.

> Note: Creating custom CPU templates requires expert knowledge of

Reviewer 2: This can be 
"> \*\*Note\*\*" (without "")
to look like:
> **Note**

(Also works with warnings)
> **Warning**

Reviewer 2: Also we need to be explicit about ARM cpu templates and that users will need patched kernel to make templates do anything useful.

Author: nice!

Author: > Also we need to be explicit about ARM cpu templates and that users will need patched kernel to make templates do anything useful.

Yes, that's a good point. Let me think how we can express that.

Author: added.

# CPU templates

Firecracker allows users to customise how the vCPUs are represented
to the guest software by changing the following configuration:

- CPUID (x86_64 only)
- MSRs (Model Specific Registers, x86_64 only)
- ARM registers (aarch64 only)

> A combination of the changes to the entities above is called a CPU template.

Reviewer 3: Nit: Grammar
```suggestion
A combination of the changes to the above entities is called a CPU template.
```

Author: done

> The functionality can be used when a user wants to mask a feature from
> the guest. One of real world use cases for this is representing a heterogeneous
> fleet (a fleet consisting of multiple CPU models) as a homogeneous fleet,
> so the guests will experience a consistent feature set supported by the host.

Reviewer 3: nit: grammar
```suggestion
The functionality can be used when a user wants to mask a feature from
the guest. A real world use case for this is representing a heterogeneous
fleet (a fleet consisting of multiple CPU models) as a homogeneous fleet,
so the guests will experience a consistent feature set supported by the host.
```

Author: done

> **Note**
Representing one CPU vendor as another CPU vendor is not supported.

> **Note**

> CPU templates is not a security feature. It does not improve
> the guest's security posture in any way. It is only supposed to modify
> indications to the guest software on what hardware features are available.

Reviewer 3: nit: We say it `does not improve the guest's security posture` but we do not say `it does not degrade the guest's security posture` we can rule out both by using `affect`.
```suggestion
CPU templates is not a security feature. It does not affect
the guest's security posture in any way. It is only supposed to modify
indications to the guest software on what hardware features are available.
```

Reviewer 3: Okay, it may just be me, but maybe this can be re-worded like:

> CPU templates is not inherently a security feature without specific configuration it does not affect a guest's security posture. When specifically configured it can however affect the security of a guest.

Author: I think I found a way to express what I actually meant. Updated.

Reviewer 3: Okay lgtm

Firecracker supports two types of CPU templates:

- Static CPU templates - a set of built-in CPU templates for users
  to choose from
- Custom CPU templates - users can create their own CPU templates in json
  format and pass them to Firecracker

> **Note**
CPU templates for ARM (both static and custom) require the following patch
to be available in the host kernel: [Support writable CPU ID registers from userspace](https://lore.kernel.org/kvm/20230212215830.2975485-1-jingzhangos@google.com/#t).
Otherwise KVM will fail to write to the ARM registers.

## Static CPU templates

> At the moment the following set of static CPU templates is supported:

Reviewer 3: nit: grammar
```suggestion
At the moment the following set of static CPU templates are supported:
```

| T2CL         | Intel      | Cascade Lake or newer |
| T2S          | Intel      | any                   |
| V1N1         | ARM        | Neoverse V1           |

T2 and C3 templates are mapped as close as possible to AWS T2 and C3 instances
in terms of CPU features.

The T2S template is designed to allow migrating [snapshots](../snapshotting/versioning.md)
between hosts with Intel Skylake and Intel Cascade Lake securely by further
restricting CPU features for the guest, however this comes with a performance
penalty. Users are encouraged to carry out a performance assessment if they wish
to use the T2S template.

The T2CL template is mapped to be close to Intel Cascade Lake.
It is not safe to use it on Intel CPUs older than Cascade Lake (such as Skylake).

The only AMD template is T2A. It is considered safe to be used with AMD Milan.

Intel T2CL and AMD T2A templates together aim to provide instruction set feature
parity between CPUs running them, so they can form a heterogeneous fleet
exposing the same instruction sets to the application.

The V1N1 template is designed to represent ARM Neoverse V1 as ARM Neoverse N1.

### Configuring static CPU templates

> Configuration of a static CPU template is performed via `/machine-config` API endpoint:

Author: done

```bash
curl --unix-socket /tmp/firecracker.socket -i  \
  -X PUT 'http://localhost/machine-config' \
  -H 'Accept: application/json'            \
  -H 'Content-Type: application/json'      \
  -d '{
           "vcpu_count": 2,
           "mem_size_mib": 1024,
           "cpu_template": "T2CL"
  }'
```

## Custom CPU templates

Users can create their own CPU templates by creating a json file containing
modifiers for CPUID, MSRs or ARM registers.

> **Note**
Creating custom CPU templates requires expert knowledge of
CPU architectures. An inappropriate configuration may lead to guest crashes or

> making guests vulnerable to security attacks. Custom CPU templates must be

Reviewer 3: This seems to conflict with the statement given earlier when it comes to the security affects of CPU templates.

This earlier statement being:

> CPU templates is not a security feature. It does not improve
the guest's security posture in any way. It is only supposed to modify

tested thoroughly before use in production.

> **Note**
Having MSRs or ARM registers in the custom CPU template does
not affect access permissions that guests will have to those registers.

> The access control is handled by KVM and is not influenced by CPU templates.

Reviewer 3: To clarify it may be helpful to add:

```
It is possible to write values in a CPU template that are not applied, this will not error.
```

In order to assist with creation and usage of CPU templates, there exists
a CPU template helper tool.
TODO: add a link to the CPU template helper tool when it is available.

### Configuring custom CPU templates

> Configuration of a custom CPU template is performed via `/cpu-config` API endpoint.

Reviewer 3: nit: grammar
```suggestion
Configuration of a custom CPU template is performed via the `/cpu-config` API endpoint.
```

Author: fixed

An example of configuring a custom CPU template on x86_64:

```bash
curl --unix-socket /tmp/firecracker.socket -i  \
  -X PUT 'http://localhost/cpu-config' \
  -H 'Accept: application/json'            \
  -H 'Content-Type: application/json'      \
  -d '{
        "cpuid_modifiers": [
          {
            "leaf": "0x1",
            "subleaf": "0x0",
            "flags": 0,
            "modifiers": [
              {
                "register": "eax",

>                 "bitmap": "0bxxxx000000000011xx00011011110010"

Reviewer 3: I don't think in reading the doc it is documented at this point that `x` passes through the value. It might be worth noting this.

Author: Thanks. I added explanation to the examples and mentioned that in the schema.

              }
            ]
          }
        ],
        "msr_modifiers": [
          {
            "addr": "0x10a",
            "bitmap": "0b0000000000000000000000000000000000000000000000000000000000000000"
          }
        ]
      }'
```

An example of configuring a custom CPU template on ARM:

```bash
curl --unix-socket /tmp/firecracker.socket -i  \
  -X PUT 'http://localhost/cpu-config' \
  -H 'Accept: application/json'            \
  -H 'Content-Type: application/json'      \
  -d '{
        "reg_modifiers": [
          {
            "addr": "0x603000000013c020",

>             "bitmap": "0bxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0000xxxxxxxxxxxx0000xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"

Reviewer 3: I don't think in reading the doc it is documented at this point that `x` passes through the value. It might be worth noting this.

not reported as supported by KVM via `KVM_GET_SUPPORTED_CPUID`,
the default register values of such a leaf are assumed to be `0`.

> #### CPUID normalization

Reviewer: It's worth to have documents about MSR modification and ARM register modification.
https://github.com/firecracker-microvm/firecracker/blob/feature/cpu-templates/src/vmm/src/vstate/vcpu/x86_64.rs#L239-L242
https://github.com/firecracker-microvm/firecracker/blob/feature/cpu-templates/src/vmm/src/vstate/vcpu/aarch64.rs#L110-L116

Author: nice, added!
